File: src/LightweightDataJson.php
<?php
/**
 * LightweightDataJson - A lightweight utility for encoding and decoding data in the .ld-json format.
 *
 * This class provides methods to encode PHP arrays into .ld-json format and decode .ld-json data into PHP arrays.
 * It can be used as open source software under the terms of the MIT License.
 * When utilizing this code, please provide appropriate attribution to the author.
 *
 * @package   LightweightDataJson
 * @category  Utility
 * @version   1.0.0
 * @license   MIT License (https://opensource.org/licenses/MIT)
 * @author    Weslley Alves
 * @link      https://github.com/kupodigital/ldjson
 */

namespace KupoDigital\Data;

class LightweightDataJson
{
    public static function decode(string $data): array
    {
        $lines = explode("\x0A", $data);
        $header = array_map(function($item) {
            return strtok($item, ":");
        }, array_map('trim', explode("\x09", trim($lines[0], "{}"))));
        $result = [];

        for ($i = 1; $i < count($lines); $i++) {
            $values = array_map('trim', explode("\x09", trim($lines[$i], "{}")));
            $object = [];

            for ($j = 0; $j < count($header); $j++) {
                if (!empty($header[$j])) {
                    $object[$header[$j]] = $values[$j];
                }
            }

            $result[] = $object;
        }

        return $result;
    }

    public static function encode(array $data): string
    {
        $header = [];
        $body = "";
        $t = 0;

        foreach ($data as $k => $value) {
            $body .= "\x0A{\x09";
            
            foreach ($value as $subKey => $subValue) {
                if ($t === 0) {
                    $header[] = "{$subKey}:" . gettype($subValue);
                }
                $body .= "{$subValue}\x09";
            }

            $t++;
            $body .= "}";
        }

        $header = implode("\x09", $header);
        return "{\x09{$header}\x09}{$body}";
    }
}

File: src/LightweightDataJson.py
class LightweightDataJson:
    
    def decode(data: str) -> list:
        lines = data.split("\n")
        header = []
        result = []

        # Encontrar a primeira linha não vazia para extrair os cabeçalhos
        for line in lines:
            if line.strip():  # Verificar se a linha não está vazia
                header = [item.split(":")[0] for item in map(str.strip, line.strip("{}").split("\t"))]
                break

        # Processar os dados após o cabeçalho
        for i in range(1, len(lines)):
            if not lines[i].strip():  # Ignorar linhas vazias
                continue
            values = list(map(str.strip, lines[i].strip("{}").split("\t")))
            obj = {}

            for j in range(len(header)):
                if header[j]:
                    obj[header[j]] = values[j]

            result.append(obj)

        return result
    
    @staticmethod
    def encode(data: list) -> str:
        header = []
        body = ""
        t = 0

        for value in data:
            body += "\x0A{\x09"
            sorted_value = dict(sorted(value.items()))
            
            for subKey, subValue in sorted_value.items():
                if t == 0:
                    header.append(f"{subKey}:{type(subValue).__name__}")
                body += f"{subValue}\x09"

            t += 1
            body += "}"

        header_str = "\x09".join(header)
        return f"{{\x09{header_str}\x09}}{body}"